### ISSUE TITLE
Lint `explicit_counter_loop` incorrectly triggers when incrementing in `let ... else ...`

### ISSUE THREAD

**Issue by laurmaedje** [rust-lang/rust-clippy#17014]
### Summary

The `explicit_counter_loop` lint seems to have a false positive when the counter is incremented in the `else` part of a `let ... else ...` statement.

### Lint Name

explicit_counter_loop

### Reproducer

I tried this code:

```rust
fn foo<T>(slice: &[Option<T>]) {
    let mut nones = 0;
    for item in slice {
        let Some(_) = item else {
            nones += 1;
            continue;
        };
    }
}
```

I saw this happen:

```
warning: the variable `nones` is used as a loop counter
 --> src/main.rs:7:5
  |
7 |     for item in slice {
  |     ^^^^^^^^^^^^^^^^^ help: consider using: `for (nones, item) in slice.iter().enumerate()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/rust-1.95.0/index.html#explicit_counter_loop
  = note: `#[warn(clippy::explicit_counter_loop)]` on by default
```

I expected to see this happen:

No warning, since the loop counter is only incremented for `None` items. The suggested change would change the semantics of the code.

### Version

```text
rustc 1.95.0 (59807616e 2026-04-14)
binary: rustc
commit-hash: 59807616e1fa2540724bfbac14d7976d7e4a3860
commit-date: 2026-04-14
host: aarch64-apple-darwin
release: 1.95.0
LLVM version: 22.1.2
```
